controller/object_store_controller.go

```diff
@@ -0,0 +1,1159 @@
+									types.ObjectStoreStatusContainerPort,
+								),
+							}, domainNameArgs...),
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "s3",
+									ContainerPort: types.ObjectStoreContainerPort,
+									Protocol:      "TCP",
+								},
+								{
+									Name:          "status",
+									ContainerPort: types.ObjectStoreStatusContainerPort,
+									Protocol:      "TCP",
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "RGW_S3GW_ENABLE_TELEMETRY",
+									Value: "true",
+								},
+							},
+							EnvFrom: env,
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      genVolumeMountName(store),
+									MountPath: "/data",
+								},
+							},
+						},
+						{
+							Name:  "s3gw-ui",
+							Image: osc.getObjectStoreUIImage(store),
+							Args:  []string{},
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "ui",
+									ContainerPort: types.ObjectStoreUIContainerPort,
+									Protocol:      "TCP",
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "S3GW_SERVICE_URL",
+									Value: fmt.Sprintf("http://127.0.0.1:%v", types.ObjectStoreContainerPort),
+								},
+								{
+									Name:  "S3GW_UI_PATH",
+									Value: fmt.Sprintf("/objectstore/%v", store.Name),
+								},
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: genVolumeMountName(store),
+							VolumeSource: corev1.VolumeSource{
+								PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+									ClaimName: genPVCName(store),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	registrySecretSetting, err := osc.ds.GetSetting(types.SettingNameRegistrySecret)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get registry secret setting for object store deployment")
+	}
+
+	if registrySecretSetting.Value != "" {
+		dpl.Spec.Template.Spec.ImagePullSecrets = []corev1.LocalObjectReference{
+			{
+				Name: registrySecretSetting.Value,
+			},
+		}
+	}
+
+	return osc.ds.CreateDeployment(&dpl)
+}
+
+// To avoid multiple longhorn managers acting on the same object store, only the
+// instance responsible for the longhorn volume is considered responsible for
+// the object store. This of course precludes that the volume has already been
+// created.
+func (osc *ObjectStoreController) isResponsibleFor(store *longhorn.ObjectStore) bool {
+	vol, err := osc.ds.GetVolumeRO(genPVName(store))
+	if err != nil && !datastore.ErrorIsNotFound(err) {
+		utilruntime.HandleError(fmt.Errorf("failed to find volume for object store %v: %v", store.Name, err))
+		return false
+	} else if err != nil {
+		return true
+	}
+
+	return osc.controllerID == vol.Status.OwnerID
+}
+
+func (osc *ObjectStoreController) getObjectStoreImage(store *longhorn.ObjectStore) string {
+	if store.Spec.Image != "" {
+		return store.Spec.Image
+	}
+	return osc.s3gwImage
```

Comment:
Should we use or check the setting value instead? Same for the below function `getObjectStoreUIImage`

Reply:
This one is already handled inside webhook. I think we should remove the check here. Inside webhook we can check and make sure that the settings `object-store-image` and `object-store-ui-image` are non-empty

Reply:
This whole code moved to the webhook. It takes care of setting the `.spec.image` and `.spec.uiImage` properties to default values read from settings if they aren't present in the API request.